stream.go

```diff
@@ -155,6 +160,11 @@ func (cs *clientStream) Trailer() metadata.MD {
 }
 
 func (cs *clientStream) SendMsg(m interface{}) (err error) {
+	cs.mu.Lock()
```

Comment:
I do not have machine to do test and it is good to know that Go mutex impl
specializes the uncontended case so that it is fast. For the "lots of
threads" case, I am not sure (it would be more subtle and depends on how
smart the compiler is and the CPU caching). But regardless, we can have a
clientStream.enableTracing boolean to gain all the sharding benefit without
having an irrelevant lock when tracing is off.

As you mentioned the resulting code is only "a bit simpler", I am not
convinced why we want to leave a hole now for future benchmark and
optimization...

On Wed, Jul 1, 2015 at 6:36 PM, David Symonds <notifications@github.com>
wrote:

> In stream.go
> <https://github.com/grpc/grpc-go/pull/235#discussion_r33741310>:
>
> > @@ -155,6 +160,11 @@ func (cs *clientStream) Trailer() metadata.MD {
> >  }
> >
> >  func (cs *clientStream) SendMsg(m interface{}) (err error) {
> > +	cs.mu.Lock()
>
> An uncontended mutex lock/unlock is a compare-and-swap (for the lock) and
> an atomic add (for the unlock), which together is around 16ns on my
> (getting old) workstation. It's definitely not "hundreds of nanoseconds".
> It's only a little slower than reading a global var (EnableTracing), which
> may, in fact, be slower if there's lots of threads (a mutex per
> clientStream is thus effectively sharding that).
>
I suggest we keep this how this change has it. It keeps the code a bit
> simpler, it's an internal only thing, and we can switch it out when higher
> speed transports arrive and 16ns is enough to show up on a benchmark.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc-go/pull/235/files#r33741310>.
>